scm/driver/gitea/milestone.go

```diff
@@ -0,0 +1,130 @@
+package gitea
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/drone/go-scm/scm"
+	"github.com/drone/go-scm/scm/driver/internal/null"
+)
+
+type milestoneService struct {
+	client *wrapper
+}
+
+func (s *milestoneService) Find(ctx context.Context, repo string, id int) (*scm.Milestone, *scm.Response, error) {
+	namespace, name := scm.Split(repo)
+	path := fmt.Sprintf("api/v1/repos/%s/%s/milestones/%d", namespace, name, id)
+	out := new(milestone)
+	res, err := s.client.do(ctx, "GET", path, nil, out)
+	return convertMilestone(out), res, err
+}
+
+func (s *milestoneService) List(ctx context.Context, repo string, opts scm.MilestoneListOptions) ([]*scm.Milestone, *scm.Response, error) {
+	namespace, name := scm.Split(repo)
+	path := fmt.Sprintf("api/v1/repos/%s/%s/milestones%s", namespace, name, encodeMilestoneListOptions(opts))
+	out := []*milestone{}
+	res, err := s.client.do(ctx, "GET", path, nil, &out)
+	return convertMilestoneList(out), res, err
+}
+
+func (s *milestoneService) Create(ctx context.Context, repo string, input *scm.MilestoneInput) (*scm.Milestone, *scm.Response, error) {
+	namespace, name := scm.Split(repo)
+	path := fmt.Sprintf("api/v1/repos/%s/%s/milestones", namespace, name)
+	in := &milestoneInput{
+		Title:       input.Title,
+		Description: input.Description,
+		State:       StateOpen,
+		Deadline:    null.NewTime(input.DueDate, true),
+	}
+	if input.State == "closed" {
+		in.State = StateClosed
+	}
+	out := new(milestone)
+	res, err := s.client.do(ctx, "POST", path, in, out)
+	return convertMilestone(out), res, err
+}
+
+func (s *milestoneService) Delete(ctx context.Context, repo string, id int) (*scm.Response, error) {
+	namespace, name := scm.Split(repo)
+	path := fmt.Sprintf("api/v1/repos/%s/%s/milestones/%d", namespace, name, id)
+	return s.client.do(ctx, "DELETE", path, nil, nil)
+}
+
+func (s *milestoneService) Update(ctx context.Context, repo string, id int, input *scm.MilestoneInput) (*scm.Milestone, *scm.Response, error) {
+	namespace, name := scm.Split(repo)
+	path := fmt.Sprintf("api/v1/repos/%s/%s/milestones/%d", namespace, name, id)
+	in := milestoneInput{}
+	if input.Title != "" {
+		in.Title = input.Title
+	}
+	switch input.State {
+	case "open":
+		in.State = StateOpen
+	case "close", "closed":
+		in.State = StateClosed
+	}
+	if input.Description != "" {
+		in.Description = input.Description
+	}
+	if !input.DueDate.IsZero() {
+		in.Deadline = null.NewTime(input.DueDate, true)
+	}
+	out := new(milestone)
+	res, err := s.client.do(ctx, "PATCH", path, in, out)
+	return convertMilestone(out), res, err
+}
+
+// StateType issue state type
+type StateType string
```

Comment:
since these types are internal to the package we should probably make sure they are private (lowercase) instead of public.

```
type stateType string
```

and

```
stateOpen stateType = "open"
```

Reply:
done